boltview/reduce.tcc

```diff
@@ -296,27 +325,32 @@ struct ReduceImplementation {
 		// constexpr int kBlockSize = 256;
 		constexpr int kBlockSize = 128;
 		constexpr int kBucketSize = 2;
-		int block_count = 1 + (size[tDimension] - 1) / (kBlockSize * kBucketSize);
 		// TODO - better setup of bucket size and block size depending on input size
-		if (block_count > 1) {
-			dim3 block(kBlockSize, 1, 1);
-			dim3 grid(block_count, reduced_size[0], reduced_size.kDimension >= 2 ? reduced_size[1] : 1);
 
-			auto tmp_size = size;
-			tmp_size[tDimension] = block_count;
-			DeviceImage<TOutputValue, TView::kDimension> tmp_buffer(tmp_size);
-			dimensionReduceKernel<TView, decltype(tmp_buffer.view()), TOutputValue, TOperator, tDimension, kBlockSize><<<grid, block, 0, execution_policy.cuda_stream>>>(inview, tmp_buffer.view(), initial_value, reduction_operator);
+		auto tmp_size = detail::estimateTemporaryBuffer(size, tDimension, kBlockSize, kBucketSize);
+		auto current_size = detail::estimateCurrentLevelTemporary(size, tDimension, kBlockSize, kBucketSize);
+		auto following_layers_offset = tmp_size - current_size;
+		auto following_layers_size = tmp_size;
+		following_layers_size[tDimension] = following_layers_offset[tDimension];
+
+		auto current_tmp = subview(tmp_view, following_layers_offset, current_size);
+		auto following_tmp = subview(tmp_view, typename TmpView::IndexType(), following_layers_size);
+
+		int block_count = current_size[tDimension];
+
+		dim3 block(kBlockSize, 1, 1);
+		dim3 grid(block_count, reduced_size[0], reduced_size.kDimension >= 2 ? reduced_size[1] : 1);
+		
+		if (block_count > 1) {
+			dimensionReduceKernel<TView, decltype(current_tmp), TOutputValue, TOperator, tDimension, kBlockSize><<<grid, block, 0, execution_policy.cuda_stream>>>(inview, current_tmp, initial_value, reduction_operator);
 			BOLT_CHECK_ERROR_AFTER_KERNEL("dimensionReduceKernel tmp buffer", grid, block);
-			run(tmp_buffer.constView(), output_view, dimension, initial_value, reduction_operator, execution_policy);
+			run(current_tmp, following_tmp, output_view, dimension, initial_value, reduction_operator, execution_policy);
 		} else {
-			dim3 block(kBlockSize, 1, 1);
-			dim3 grid(1 + (size[tDimension] - 1) / block.x, reduced_size[0], reduced_size.kDimension >= 2 ? reduced_size[1] : 1);
```

Comment:
Why was the x-coordinate of grid size calculated differently in this case?

Reply:
seems like a bug - it should be 1 (or block_count)